File: web/backend/api/store/catalog.go
package store

import (
	"encoding/json"
	"net/http"
)

// Basic product catalog (what we originally planned)

func (h *Handler) handleListProducts(w http.ResponseWriter, r *http.Request) {
	productType := r.URL.Query().Get("type")
	// TODO: query aoStore for products, filter by type if provided
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]any{
		"products": []map[string]any{
			{
				"id":          "device-123",
				"name":        "Web Search Skill",
				"description": "A skill that searches the web",
				"type":        "Device",
				"version":     "1.0.0",
			},
		},
	})
}

func (h *Handler) handleGetProduct(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	// TODO: fetch product details from aoStore
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]any{
		"id":          id,
		"name":        "Example Product",
		"description": "Detailed description of the product",
		"type":        "Device",
		"metadata":    map[string]any{},
	})
}

func (h *Handler) handleInstallProduct(w http.ResponseWriter, r *http.Request) {
	var req struct {
		ProductID string `json:"product_id"`
		Version   string `json:"version,omitempty"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "invalid request body", http.StatusBadRequest)
		return
	}

	// TODO: download and install the skill, update workspace

	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(map[string]string{
		"status": "installed",
		"path":   "/path/to/installed/skill", // placeholder
	})
}

File: web/frontend/src/api/store/catalog.ts
// Catalog API – product listing, details, and installation

export interface Product {
  id: string
  name: string
  description: string
  type: string          // "Device", "DApp", "Project"
  version?: string
  metadata?: Record<string, any>
}

export interface ProductsResponse {
  products: Product[]
}

export interface ProductDetailResponse extends Product {}

export interface InstallRequest {
  product_id: string
  version?: string
}

export interface InstallResponse {
  status: string
  path: string
}

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const res = await fetch(path, options)
  if (!res.ok) {
    let message = `API error: ${res.status} ${res.statusText}`
    try {
      const body = (await res.json()) as {
        error?: string
        errors?: string[]
      }
      if (Array.isArray(body.errors) && body.errors.length > 0) {
        message = body.errors.join('; ')
      } else if (typeof body.error === 'string' && body.error.trim() !== '') {
        message = body.error
      }
    } catch {
      // ignore invalid body
    }
    throw new Error(message)
  }
  return res.json() as Promise<T>
}

export async function listProducts(type?: string): Promise<ProductsResponse> {
  const url = type ? `/api/store/products?type=${encodeURIComponent(type)}` : '/api/store/products'
  return request<ProductsResponse>(url)
}

export async function getProduct(id: string): Promise<ProductDetailResponse> {
  return request<ProductDetailResponse>(`/api/store/products/${encodeURIComponent(id)}`)
}

export async function installProduct(req: InstallRequest): Promise<InstallResponse> {
  return request<InstallResponse>('/api/store/install', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
  })
}